v2/metadata/client.go

```diff
@@ -768,6 +788,77 @@ func (c *Client) RecordArtifact(ctx context.Context, outputName, schema string,
 	}, nil
 }
 
+func (c *Client) GetOrInsertArtifactType(ctx context.Context, schema string) (typeID int64, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("getOrInsertArtifactType(schema=%q) failed: %w", schema, err)
+		}
+	}()
+	at, err := SchemaToArtifactType(schema)
+	if err != nil {
+		return 0, err
+	}
+	getTypesRes, err := c.svc.GetArtifactType(ctx, &pb.GetArtifactTypeRequest{TypeName: at.Name})
+	if err != nil {
+		return 0, err
+	}
+	if getTypesRes.GetArtifactType() != nil {
+		return getTypesRes.GetArtifactType().GetId(), nil
+	}
+	putTypeRes, err := c.svc.PutArtifactType(ctx, &pb.PutArtifactTypeRequest{ArtifactType: at})
+	if err != nil {
+		return 0, err
+	}
+	return putTypeRes.GetTypeId(), err
+
+}
+
+func (c *Client) FindMatchedArtifact(ctx context.Context, artifactToMatch *pb.Artifact, pipelineContextId int64) (matchedArtifact *pb.Artifact, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("FindMatchedArtifact(artifact=%q) failed: %w", artifactToMatch, err)
+		}
+	}()
+	uris := []string{artifactToMatch.GetUri()}
+	getArtifactsByUriRes, err := c.svc.GetArtifactsByURI(ctx, &pb.GetArtifactsByURIRequest{Uris: uris})
+	if err != nil {
+		return nil, err
+	}
+	for _, candidateArtifact := range getArtifactsByUriRes.GetArtifacts() {
+		matched, err := c.matchedArtifactOrNot(ctx, artifactToMatch, candidateArtifact, pipelineContextId)
+		if err != nil {
+			return nil, err
+		}
+		if matched {
+			return candidateArtifact, nil
+		}
+	}
+	return nil, nil
+
+}
+
+func (c *Client) matchedArtifactOrNot(ctx context.Context, target *pb.Artifact, candidate *pb.Artifact, pipelineContextId int64) (bool, error) {
+	if target.GetTypeId() != candidate.GetTypeId() || target.GetState() != candidate.GetState() || target.GetUri() != candidate.GetUri() {
+		return false, nil
+	}
+	for target_k, target_v := range target.GetCustomProperties() {
+		val, ok := candidate.GetCustomProperties()[target_k]
+		if !ok || !proto.Equal(target_v, val) {
+			return false, nil
+		}
+	}
+	getConextsByArtifactRes, err := c.svc.GetContextsByArtifact(ctx, &pb.GetContextsByArtifactRequest{ArtifactId: candidate.Id})
```

Comment:
nit: typo `getContexts`

note, the variable is only used in a very small scope, so it's clear enough just naming it `res` I think.
Google internal ref: go/go-style/decisions#variable-names

Reply:
done